Do not return error for extended SNS / SQS attribute types

AWS allows using custom attribute types, like "Number.float", which is documented for SQS (https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-message-metadata.html#message-attribute-data-types).

It's not mentioned in SNS documentation (https://docs.aws.amazon.com/sns/latest/dg/sns-message-attributes.html#SNSMessageAttributes.DataTypes) but it works in production for a long time while we use Spring Cloud AWS, which performs the following operation: https://github.com/awspring/spring-cloud-aws/blob/v3.4.2/spring-cloud-aws-sns/src/main/java/io/awspring/cloud/sns/core/TopicMessageChannel.java#L106 - which results in "timestamp" attribute having "Number.java.lang.Long" type.

When we pointed the application to Moto, we got such an exception:
```
org.springframework.messaging.MessageDeliveryException: Failed to send message to TopicMessageChannel[TopicMessageChannel@5467cc55], failedMessage=GenericMessage [payload={ not important here }, headers={id=a98a7d74-d662-0dd5-4bee-a84068e30a42, contentType=application/json, timestamp=1772124243491}]
	at org.springframework.messaging.support.AbstractMessageChannel.send(AbstractMessageChannel.java:149) ~[spring-messaging-6.2.16.jar:6.2.16]
	at org.springframework.messaging.support.AbstractMessageChannel.send(AbstractMessageChannel.java:125) ~[spring-messaging-6.2.16.jar:6.2.16]
	at io.awspring.cloud.sns.core.SnsTemplate.doSend(SnsTemplate.java:83) ~[spring-cloud-aws-sns-3.4.2.jar:3.4.2]
	at io.awspring.cloud.sns.core.SnsTemplate.doSend(SnsTemplate.java:49) ~[spring-cloud-aws-sns-3.4.2.jar:3.4.2]
	at org.springframework.messaging.core.AbstractMessageSendingTemplate.send(AbstractMessageSendingTemplate.java:109) ~[spring-messaging-6.2.16.jar:6.2.16]
	at org.springframework.messaging.core.AbstractMessageSendingTemplate.convertAndSend(AbstractMessageSendingTemplate.java:148) ~[spring-messaging-6.2.16.jar:6.2.16]
	at io.awspring.cloud.sns.core.SnsTemplate.convertAndSend(SnsTemplate.java:111) ~[spring-cloud-aws-sns-3.4.2.jar:3.4.2]
	at io.awspring.cloud.sns.core.SnsTemplate.convertAndSend(SnsTemplate.java:99) ~[spring-cloud-aws-sns-3.4.2.jar:3.4.2]
...
Caused by: software.amazon.awssdk.services.sns.model.InvalidParameterException: The message attribute 'timestamp' has an invalid message attribute type, the set of supported type prefixes is Binary, Number, and String. (Service: Sns, Status Code: 400, Request ID: not important here) (SDK Attempt Count: 1)
```

We observe the issue only for SNS. Spring Cloud AWS has similar logic for SQS (https://github.com/awspring/spring-cloud-aws/blob/v3.4.2/spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/support/converter/SqsHeaderMapper.java#L146) but "timestamp" attribute is skipped.

We'd like that extended SNS / SQS attribute types do not result in an error. That's the main blocker for us to use Moto.

Thank you for the work you're doing.

Moto version: Docker image `motoserver/moto:5.1.21`

--- a/moto/sns/responses.py
+++ b/moto/sns/responses.py
@@ -84,7 +84,7 @@ class SNSResponse(BaseResponse):
                 )
 
             data_type_parts = data_type.split(".")
-            if len(data_type_parts) > 2 or data_type_parts[0] not in [
+            if data_type_parts[0] not in [
                 "String",
                 "Binary",
                 "Number",

--- a/tests/test_sns/test_publishing.py
+++ b/tests/test_sns/test_publishing.py
@@ -300,6 +300,41 @@ def test_publish_to_sqs_msg_attr_byte_value():
 
 
 @mock_aws
+def test_publish_to_sqs_with_custom_data_type():
+    conn = boto3.client("sns", region_name="us-east-1")
+    conn.create_topic(Name="some-topic")
+    response = conn.list_topics()
+    topic_arn = response["Topics"][0]["TopicArn"]
+    sqs = boto3.resource("sqs", region_name="us-east-1")
+    queue = sqs.create_queue(QueueName="test-queue")
+    conn.subscribe(
+        TopicArn=topic_arn, Protocol="sqs", Endpoint=queue.attributes["QueueArn"]
+    )
+    queue_raw = sqs.create_queue(QueueName="test-queue-raw")
+    conn.subscribe(
+        TopicArn=topic_arn,
+        Protocol="sqs",
+        Endpoint=queue_raw.attributes["QueueArn"],
+        Attributes={"RawMessageDelivery": "true"},
+    )
+    conn.publish(
+        TopicArn=topic_arn,
+        Message="my message",
+        MessageAttributes={
+            "store": {"DataType": "Number.java.lang.Long", "StringValue": "42"}
+        },
+    )
+    message = json.loads(queue.receive_messages()[0].body)
+    assert message["Message"] == "my message"
+    assert message["MessageAttributes"] == {
+        "store": {
+            "Type": "Number.java.lang.Long",
+            "Value": "42",
+        }
+    }
+
+
+@mock_aws
 def test_publish_to_sqs_msg_attr_number_type():
     sns = boto3.resource("sns", region_name="us-east-1")
     topic = sns.create_topic(Name="test-topic")

